keps/sig-api-machinery/20190226-network-proxy.md

```diff
@@ -0,0 +1,510 @@
+The proxy should not require the NetworkContext. This means the service CIDR,
+node CIDR and pod CIDR of each cluster network cannot overlap.
+
+### Network Context
+
+The minimal NetworkContext looks like the following struct in golang:
+
+```go
+type NetworkContext struct {
+        // NetworkName is the unique name of the network
+        // to select to route the traffic to.
+        NetworkName string
+}
+```
+
+NetworkName specifies the network to route traffic to.
+The KAS starts with a list of registered network names and
+the corresponding Connectivity services. So the KAS knows where to
+proxy the traffic to, otherwise it return an “Unknown network” error.
+
+The KAS starts with a proxy configuration like the below example.
+The example specifies 4 networks. "direct" specifies the KAS talking directly
+on the local network (no proxy). "master" specifies the KAS talks to a proxy
+listening at 1.2.3.4:5678. "cluster" specifies the KAS talk to a proxy
+listening at 1.2.3.5:5679. While these are represented as resources
+they are not intended to be loaded dynamically.
+The KAS loads them as configuration at start time.
+
+```yaml
+apiVersion: connectivityservice.k8s.io/v1alpha1
+kind: ConnectivityServiceConfiguration
+connectionService:
+  name: direct
+  connection:
+    type: direct
+---
+apiVersion: connectivityservice.k8s.io/v1alpha1
+kind: ConnectivityServiceConfiguration
+connectionService:
+  name: Master
+  connection:
+    type: grpc
+    url: grpc://1.2.3.4:5678
+    caBundle: file1.pem
+    clientKeyFile: proxy-client1.key
+    clientCertFile: proxy-client1.crt
+---
+apiVersion: connectivityservice.k8s.io/v1alpha1
+kind: ConnectivityServiceConfiguration
+connectionService:
+  name: Cluster
+  connection:
+    type: grpc
+    url: grpc://1.2.3.5:5679
+    caBundle: file2.pem
+    clientKeyFile: proxy-client2.key
+    clientCertFile: proxy-client2.crt
+```
+
+NetworkContext could be extended to contain more contextual information.
+This would allow smarter routing based on the k8s object KAS is processing
+or which user/tenant tries to initiate the request, etc.
+
+### Proxy gRPC definition
+
+In order to serve a proxy request, one gRPC bidirectional stream on proxy
+server is created to serve it. It's a 1:1 mapping from TCP connection to a
+gRPC stream, so the state of TCP connection is exactly the same as the gRPC
+stream state.
+
+```grpc
+syntax = "proto3";
+
+service ProxyService {
+  // Proxy a TCP connection to a remote address defined by ConnectParam.
+  // The ConnectParam is defined in metadata under key "x-kube-net-proxy".
+  // metadata["x-kube-net-proxy"] = base64.Encode(proto.Marshal(connectOptions))
+  rpc Proxy(stream Payload) returns (stream Payload) {}
+}
+
+// ConnectOptions defines the remote TCP endpoint to connect
+message ConnectOptions {
+  string remote_addr = 1; // remote address to connect to. e.g. 8.8.8.8:53
+}
+
+// Payload defines a TCP payload.
+message Payload {
+  bytes data = 1;
```

Comment:
we don't have any provision for errors?

Reply:
Working through this item. We expect to have more details on this later.